### PR TITLE
eXtensible Markup Language

### PR DESCRIPTION
**Pull request by kmeisthax**
Flash contains an internal XML parser, DOM, and object hierarchy for AVM1 that we need to support in order for XML-driven menus (i.e. much of Homestar Runner's category pages) and other content to work. This PR implements a DOM and AS2 object hierarchy for parsing XML.

## Progress

* [x] `XMLNode` & `XMLNode.prototype`
  * [x] `new XMLNode()` constructor
  * [x] `appendChild()`
  * [x] `cloneNode()`
  * [x] `getNamespaceForPrefix()`
  * [x] `getPrefixForNamespace()`
  * [x] `hasChildNodes()`
  * [x] `insertBefore()`
  * [x] `removeNode()`
  * [x] `toString()`
  * [x] `attributes`
  * [x] `childNodes`
  * [x] `firstChild`
  * [x] `lastChild`
  * [x] `localName`
  * [x] `namespaceURI`
  * [x] `nextSibling`
  * [x] `nodeName`
  * [x] `nodeType`
  * [x] `nodeValue`
  * [x] `parentNode`
  * [x] `prefix`
  * [x] `previousSibling`
* [x] `XML` & `XML.prototype`
  * [x] `new XML()` constructor
  * [x] Parsing strings into XML with `new XML`
  * [x] `docTypeDecl`
  * [x] `idMap`
  * [x] `status` (trivial impl)
  * [x] ~`xmlDecl`~
    * The `xmlDecl` in AVM1's XML handling facility doesn't actually validate XML declarations, but our parser does. In order to get around this, we need to hack or modify our parser to allow access to the exact text contents that were parsed.
    * Since I'm moving that particular bit of `xmlDecl` out of scope, I should note that reading `xmlDecl`s is otherwise fine, but it is different enough in such a way that tests will fail.
  * [x] `createElement`
  * [x] `createTextNode`
  * [x] `parseXML`
* [ ] SWF version gates for all of the above

## What's not in scope?
Features that require the ability to fetch data. Properties and methods that rely upon them have been excluded from the list above. We won't be able to implement them until after `moviefetch` is merged in (see PR #120 for more info).

## Wait, why a custom DOM?
We use a custom DOM implementation built on top of `quick_xml`'s event-driven parser. Existing implementations of XML DOM in Rust were rejected fairly quickly due to a combination of requirements:

1. XML objects need to be able to hold individual DOM nodes.
2. XML objects need to be able to navigate the DOM tree in all directions.
3. DOM nodes need to be able to hold XML objects in order for expando properties to persist correctly.

Requirement 1 is a problem because most DOM implementations in Rust use tree-structured ownership as their memory management strategy. In other words, each DOM node is directly owned by the parent - either `Element` directly contains an `Vec<Element>`, or contains `Vec<Node>` where `Node` is an enum that contains an `Element` variant. This precludes efficiently referencing child nodes from a garbage-collected object with `'gc` lifetime - we would have to hold the whole tree in an `Rc` and then traverse it any time we want to interact with it.

Requirement 2 is a further complication upon the first. Since most Rust DOMs are tree-owned, you can't have sibling or parent references that make any sort of sense. In order to avoid this, you'd need to have each node own `Rc`s to their children and hold `Weak` references to siblings and parents. If such a DOM exists, we could have used it and held node `Rc`s in `XMLObject` instead... except,

Requirement 3 makes even that impractical. In Flash, you can stick expando properties on XML DOM nodes, and they'll persist forever as if it was a native object. So our own XML DOM needs to be able to retain AVM object data. This requires, at the minimum, an XML DOM that can hold an arbitrary `T` per node; and practically must also derive `Collect` in order to work with `GcCell`s. I was not able to find any DOM implementation that does this.

As a result, this PR ships with a custom DOM that is exclusively designed to work with Ruffle. It's completely garbage collected, and every XML node can optionally be linked to an AVM object, just like `DisplayObject`. This is in the `xml` module and it is intended to be the primary representation of all XML data in Ruffle going forward.

### CODE REVIEW

**Comment by kmeisthax** [comment]
So, when implementing the version gates for XML nodes, I discovered that a significant number of player versions listed in AS2LR are just flat-out incorrect, for *several* reasons. First off, some of them were implemented and accessible in a prior Flash Player version, but not documented by Macromedia. Second, Macromedia did not implement version gates for *any* XML properties that I'm aware of. That is to say, if you access `XMLNode.prefix` in a file targeting Flash Player 5, you will get the same value as if you had targeted Flash Player 8, but `undefined` if your file is *actually run* in Flash Player 5.

Based on this, I don't believe it's appropriate to implement *any* version-specific behavior for `XML` or `XMLNode` at this time. I believe this PR is ready for review and merge.

File: core/src/avm1/globals/xml.rs [hunk 1]
```diff
@@ -0,0 +1,801 @@
...
+    _args: &[Value<'gc>],
+) -> Result<ReturnValue<'gc>, Error> {
+    if let Some(xmlnode) = this.as_xml_node() {
+        Ok((xmlnode.children_len() > 0).into())
+    } else {
+        Ok(Value::Undefined.into())
+    }
+}
+
+#[allow(unused_must_use)]
```
[the remark this answers is not in the document]
Comment (kmeisthax) [reply core/src/avm1/globals/xml.rs]: These were development leftovers, I have removed them (or will when I finish going through these comments and push the changes). My current strategy for any method that yields an `Err` is to log it and return; removing the `#[allow]` actually caught an issue where I had forgotten to do that. Thanks.

**Comment by kmeisthax** [comment]
So, `attributes` is a particularly unique case: writing any string value to it will actually construct a new attribute on the element it's attached to with that value as a string. We don't support `__resolve` yet, and the real `attributes` doesn't have one, so the most straightforward way to handle it was a native object type.

(Interesting but useless fact: you can set `__proto__` on attributes. If you do this, the underlying element gets a `__proto__` attribute set to `[object Object]`, *and* the attributes object becomes an object of that type.)

`idMap` isn't as crazy as `attributes`, but in order for it to be a regular object I would need to either construct every `XMLObject` at time of creation of the `idMap` (which seemed wasteful), or add virtual properties for each property (which wouldn't work), or use `__resolve` (which we don't have).

It may make more sense to have multiple variants of `XMLObject` rather than three different object types with slightly different `get_local` impls. I'll let you make the final decision on what I should do, though.

**Comment by Herschel** [comment]
What I mean is that we don't even create the actual XML nodes until actual XML output is requested via `toString` or etc. -- so then attributes and everything is just a dumb object, and we enumerate it when we need to create the actual XML. (Just thinking aloud, because I had some ideas to simplify Object in general -- I could elaborate more on Discord.)

More importantly, I like getting more things up and running. This is great, thanks for your work!

**Pull request merged**